<br><br>

<center><u><font size="5" color="blue">Probability Question</font></u></center>


<br>



`Calculate the probability that 5 independent random selections of an integer from 
the set of integers {1, 2, 3, 4, 5} will result in three 4s and two 5s in any order. `

<br>

<b>We will approach this first through combinatrics...</b>

<br><br>

<pre>
			
			This is 5 choose 5 with replacement
            Order doesnt matter, that is {4,4,4,5,5} and {5,4,5,4,4} are both success
            
</pre>
            

<br>

First use itertools.product to create a <i>Cartesian Product</i> to get all possible permutaitons.

<br>

In [1]:
from itertools import product 
total_perms=len(list(product([1,2,3,4,5], repeat=5))) 
print("There are ", total_perms, " total permutations if we select 5 times with replacement")

There are  3125  total permutations if we select 5 times with replacement


<br>

Then get the number of {4,4,4,5,5} sorts. 

This is tricky because we have one number repeating 3 times and another number repeating 2 times.

We dont want the total permutations of five numbers, just the 5 sorts out of the 3125 total permutations.

We can do this by factoring out 3! and 2! from 5!

<br>

In [6]:
import math

total_sorts=math.factorial(5)/(math.factorial(3)*math.factorial(2))

print("There are ", total_sorts, " total sorts from 3 fours and 2 fives")


There are  10.0  total sorts from 3 fours and 2 fives


In [7]:

print("The probability then is ", total_sorts/total_perms)


The probability then is  0.0032


<br>

<b>Now we will use simultation to help validate our combinatric solution...</b>

<br>


In [9]:

success=0
fail=0
n=1000000

for x in range(n):
    res=np.random.choice(a=[1,2,3,4,5], size =5)
    fours=sum(1 for i in res if i == 4)
    fives=sum(1 for i in res if i == 5)
    if fours==3  and fives==2:
        success=success+1
    else:
        fail=fail+1
    
print(success, " success and ", fail, " fails. Probability = " , success/n)

3231  success and  996769  fails. Probability =  0.003231


<br>

<b>One more method...loop through the products and just count the number of 3 fours and 2 fives</b>

<br>

In [10]:
success=0
fail=0
for res in product([1,2,3,4,5], repeat=5):
    if res.count(4)==3  and res.count(5)==2:
        success=success+1
    else:
        fail=fail+1
    if (res.count(5)==5):
        print("All fives")
    if (res.count(4)==5):
        print("All fours")        

print("Success =" , success, " and Faile = ", fail, " = " , success/(success+fail))


All fours
All fives
Success = 10  and Faile =  3115  =  0.0032
